agixt/Memories.py

```diff
@@ -39,110 +48,104 @@ def load_spacy_model(self):
-                    chroma_db_impl="duckdb+parquet",
-                    persist_directory=self.chroma_persist_dir,
-                    anonymized_telemetry=False,
-                )
+            memories_exist = await self.chroma_client.does_collection_exist_async(
+                "memories"
             )
+            if not memories_exist:
+                await self.chroma_client.create_collection_async(
+                    collection_name="memories"
+                )
+                memories = await self.chroma_client.get_collection_async(
+                    collection_name="memories"
+                )
+            else:
+                memories = await self.chroma_client.get_collection_async(
+                    collection_name="memories"
+                )
+            return memories
         except Exception as e:
             raise RuntimeError(f"Unable to initialize chroma client: {e}")
 
-    def get_or_create_collection(self):
-        if not self.chroma_client:
-            self.chroma_client = self.initialize_chroma_client()
-        embedder = Embedding(self.agent_config)
-        self.embedding_function = embedder.embed
-        self.chunk_size = embedder.chunk_size
-        try:
-            return self.chroma_client.get_collection(
-                name="memories", embedding_function=self.embedding_function
-            )
-        except ValueError:
-            logging.info(f"Memories for {self.agent_name} do not exist. Creating...")
-            return self.chroma_client.create_collection(
-                name="memories", embedding_function=self.embedding_function
-            )
-
-    def generate_id(self, content: str, timestamp: str):
-        return sha256((content + timestamp).encode()).hexdigest()
+    async def store_memory(
+        self, content: str, description: str = None, external_source_name: str = None
+    ):
+        embedder, chunk_size = await self.get_embedder()
+        collection = await self.get_collection()
+        record = MemoryRecord(
+            is_reference=False,
+            id=sha256((content + datetime.now().isoformat()).encode()).hexdigest(),
+            text=content,
+            timestamp=datetime.now().isoformat(),
+            description=description,
+            external_source_name=external_source_name,  # URL or File path
+            embedding=await embedder(content),
+        )
 
-    def store_memory(self, id: str, content: str, metadatas: dict):
-        if not self.chroma_client:
-            self.chroma_client = self.initialize_chroma_client()
-            self.collection = self.get_or_create_collection()
         try:
-            self.collection.add(
-                ids=id,
-                documents=content,
-                metadatas=metadatas,
+            await self.chroma_client.upsert_async(
+                collection_name="memories",
+                record=record,
             )
+            self.chroma_client._client.persist()
```

Comment:
@Josh-XT

Reply:
It isn't an async function, persist is basically hitting "save" from what I understand.